plonky2/src/fri/batch_prover.rs

```diff
@@ -0,0 +1,457 @@
+        timing,
+        "fold codewords in the commitment phase",
+        batch_fri_committed_trees::<F, C, D>(
+            lde_polynomial_coeffs,
+            lde_polynomial_values,
+            challenger,
+            fri_params,
+        )
+    );
+
+    // PoW phase
+    let pow_witness = timed!(
+        timing,
+        "find proof-of-work witness",
+        fri_proof_of_work::<F, C, D>(challenger, &fri_params.config)
+    );
+
+    // Query phase
+    let query_round_proofs = batch_fri_prover_query_rounds::<F, C, D>(
+        initial_merkle_trees,
+        &trees,
+        challenger,
+        n,
+        fri_params,
+    );
+
+    FriProof {
+        commit_phase_merkle_caps: trees.iter().map(|t| t.cap.clone()).collect(),
+        query_round_proofs,
+        final_poly: final_coeffs,
+        pow_witness,
+    }
+}
+
+pub(crate) fn batch_fri_committed_trees<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+>(
+    mut final_coeffs: PolynomialCoeffs<F::Extension>,
+    values: &[PolynomialValues<F::Extension>],
+    challenger: &mut Challenger<F, C::Hasher>,
+    fri_params: &FriParams,
+) -> FriCommitedTrees<F, C, D> {
+    let mut trees = Vec::with_capacity(fri_params.reduction_arity_bits.len());
+    let mut shift = F::MULTIPLICATIVE_GROUP_GENERATOR;
+    let mut polynomial_index = 1;
+    let mut final_values = values[0].clone();
+    for arity_bits in &fri_params.reduction_arity_bits {
+        let arity = 1 << arity_bits;
+
+        reverse_index_bits_in_place(&mut final_values.values);
+        let chunked_values = final_values.values.par_chunks(arity).map(flatten).collect();
+        let tree = MerkleTree::<F, C::Hasher>::new(chunked_values, fri_params.config.cap_height);
+
+        challenger.observe_cap(&tree.cap);
+        trees.push(tree);
+
+        let beta = challenger.get_extension_challenge::<D>();
+        // P(x) = sum_{i<r} x^i * P_i(x^r) becomes sum_{i<r} beta^i * P_i(x).
+        // TODO: Optimize the folding process. Consider folding the functions directly in the value domain.
+        final_coeffs = PolynomialCoeffs::new(
+            final_coeffs
+                .coeffs
+                .par_chunks_exact(arity)
+                .map(|chunk| reduce_with_powers(chunk, beta))
+                .collect::<Vec<_>>(),
+        );
+        shift = shift.exp_u64(arity as u64);
+        final_values = final_coeffs.coset_fft(shift.into());
+        if polynomial_index != values.len() && final_values.len() == values[polynomial_index].len()
+        {
+            final_values = PolynomialValues::new(
+                final_values
+                    .values
+                    .iter()
+                    .zip(&values[polynomial_index].values)
+                    .map(|(&f, &v)| f * beta + v)
+                    .collect::<Vec<_>>(),
+            );
+            polynomial_index += 1;
+        }
+        final_coeffs = final_values.clone().coset_ifft(shift.into());
```

Comment:
I think the folding process here is convoluted because after a round of FRI reduction, we naturally pass from domain $g.H$ to $g^2H^2$. I am still not exactly able to figure out why exactly this works, but I am unsure how `f`(in domain $g^2H^2$ is able to talk with `v` (in domain $gH^2$). There ought to be some bug here.
Also, beta ought not be the challenge with which we should be able to combine the polynomials(since it is different each round). Hence `f * beta + v` sounds incorrect.

I have attempted to fix  this in https://github.com/0xmozak/plonky2/pull/91. Basically it ensures that polynomials in a batch are evaluated with coset shift given by `F::gen().exp(log_max_degree -log_d)`. Where `log_max_degree` is log of max degree among polynomials in the batch. This ensures that after a round of FRI reduction, the values remain in their natural domain. More explanation in the PR

Reply:
Thanks for the review!

Could you please explain more why 'f * beta + v' sounds incorrect if beta varies in each round?

Reply:
Ok incorrect might not be the right word, but ideally I thought we should have the random challenge right after we commit to all polynomials in the batch, not after intermediate FRI rounds too. I have added description in the PR, which might make my reasoning more clear.

Reply:
Thanks for the detailed explanation in #91!
I think your explanation has no problems. However, I would like to understand if the current design is indeed incorrect.

Instead of running the FRI protocol with a random combination of the polynomials

$p_1(x) + \alpha p_2(x) + \alpha^2 p_3(x) \cdots$

the current design is like running the FRI protocol individually on

$p_1(x), p_2(x), p_3(x) ...$

We just fold in the polynomials $p_2(x),p_3(x),...$ into the FRI protocol of $p_1(x)$ at certain degrees. It seems not to reduce any security assumptions introduced by FRI.

Reply:
I think the presentation in the code here makes the flow of information a bit harder to understand.

We can shift perspective.  In the current code we have `f := reduce_with_powers(chunk, beta)` and `f * beta + v`.

Equivalently, look at `f := beta * reduce_with_powers(chunk, beta)` and `f + v`.

Just think of replace `reduce_with_powers` with a function that mangles `chunk` more effectively (by increasing all powers of `beta` by one), so that the malicious prover can not predict anything about hitting any particular values.  (Whereas when you only do a pure `reduce_with_powers`, then a malicious prover gets direct influence over `f` via the zeroth entry in each chunk.)

That shift of perspective works just as well in your branch, @codeblooded1729.

---

In plonky3, they have a functions to work with 'shifted powers', that don't start at 1, but at some other value.  

In general, I also think much of our code would benefit from shifted powers not iterating backwards, but forwards instead.  (Only the code, the math would stay the same or equivalent.) But that's another discussion entirely.